javalib/src/main/scala/java/util/regex/PatternCompiler.scala

```diff
@@ -0,0 +1,1759 @@
+   * listed here. In addition, we have 0009-000D and 0085 from the Cc category.
+   * Therefore, the following equivalence holds:
+   *
+   *   \p{IsWhite_Space} === [\x09-\x0d\x85\p{gc=Zs}\p{gc=Zl}\p{gc=Zp}]
+   *
+   * That equivalence is known to be true as of Unicode 13.0.0, and seems to
+   * have been true for a number of past versions as well. We rely on it to
+   * define \p{Print} and \p{Blank} above. Those would become buggy if a future
+   * version of Unicode invalidates that assumption.
+   */
+
+  private val scriptCanonicalizeRegExp = new js.RegExp("(?:^|_)[a-z]", "g")
+
+  /** Canonicalizes a script name's casing.
+   *
+   *  The JDK regexps compare script names while ignoring case, but JavaScript
+   *  requires the canonical name.
+   */
+  private def canonicalizeScriptName(scriptName: String): String = {
+    import js.JSStringOps._
+
+    val lowercase = scriptName.toLowerCase()
+
+    if (lowercase == "signwriting") {
+      // Exception: uppercase 'W' even though not after '_'
+      "SignWriting"
+    } else {
+      lowercase.jsReplace(scriptCanonicalizeRegExp,
+          ((s: String) => s.toUpperCase()): js.Function1[String, String])
+    }
+  }
+
+  @inline
+  private final class CodePointRange(val start: Int, val end: Int) {
+    def isEmpty: Boolean = start > end
+    def nonEmpty: Boolean = start <= end
+
+    /** Computes the intersection of two *non-empty* ranges.
+     *
+     *  This method makes no guarantee about its result if either or both input
+     *  ranges are empty.
+     *
+     *  The result range may be empty.
+     */
+    def intersect(that: CodePointRange): CodePointRange =
+      CodePointRange(Math.max(this.start, that.start), Math.min(this.end, that.end))
+
+    def shift(offset: Int): CodePointRange =
+      CodePointRange(start + offset, end + offset)
+  }
+
+  private object CodePointRange {
+    @inline
+    def apply(start: Int, end: Int): CodePointRange =
+      new CodePointRange(start, end)
+
+    @inline
+    def BmpBelowHighSurrogates: CodePointRange =
+      CodePointRange(0, Character.MIN_HIGH_SURROGATE - 1)
+
+    @inline
+    def HighSurrogates: CodePointRange =
+      CodePointRange(Character.MIN_HIGH_SURROGATE, Character.MAX_HIGH_SURROGATE)
+
+    @inline
+    def BmpAboveHighSurrogates: CodePointRange =
+      CodePointRange(Character.MAX_HIGH_SURROGATE + 1, Character.MAX_VALUE)
+
+    @inline
+    def Supplementaries: CodePointRange =
+      CodePointRange(Character.MIN_SUPPLEMENTARY_CODE_POINT, Character.MAX_CODE_POINT)
+  }
+
+  private final class CharacterClassBuilder(asciiCaseInsensitive: Boolean, isNegated: Boolean) {
+    private var conjunction = ""
+    private var thisConjunct = ""
+    private var thisSegment = ""
+
+    def finish(): String = {
+      val conjunct = conjunctResult()
+      if (conjunction == "") conjunct else s"(?:$conjunction$conjunct)"
+    }
+
+    def finishConjunct(): Unit = {
```

Comment:
I think this is misnamed: It doesn't *finish* the conjunct, it starts a new one. Maybe call it `beginConjunct`? (also not 100% accurate, I realize, because it need not be called for a single conjunct :-/)

Reply:
`finishConjunctAndStartNewOne()`? Or `finishNonLastConjunct()`? Or `startNewConjunct()` (the "new" suggesting that it's not the first one)?